Interface injection chooses an overly broad dependency set

I have a project where my main source set has a limited subset of fabric api modules on the compile classpath, and a separate testmod source set has all of fabric API available at compile- and runtime.

Currently interface injection sums all remapped configurations together, which injects unresolvable classes onto the compile classpath for some configuration, producing errors like:

```
C:\Users\abby\dev\kyori-text\adventure-platform-fabric\src\main\java\net\kyori\adventure\platform\fabric\impl\server\ServerPlayerAudience.java:236: error: cannot access FabricItem
    final ItemStack bookStack = new ItemStack(Items.WRITTEN_BOOK, 1);
                                ^
  class file for net.fabricmc.fabric.api.item.v1.FabricItem not found
C:\Users\abby\dev\kyori-text\adventure-platform-fabric\src\main\java\net\kyori\adventure\platform\fabric\impl\server\ServerPlayerAudience.java:237: error: cannot access FabricItemStack
    final CompoundTag bookTag = bookStack.getOrCreateTag();
                                         ^
  class file for net.fabricmc.fabric.api.item.v1.FabricItemStack not found
```

Is there a sensible fully automated default here? My feeling is that this may require adding additional configurability to the interface injection extension, maybe taking a list of remapped configurations to use for injection?

I've [temporarily worked around this issue](https://github.com/KyoriPowered/adventure-platform-fabric/blob/mc/1.19/build.gradle.kts#L143) by disabling dependency interface injection entirely.

--- a/src/main/java/net/fabricmc/loom/api/RemapConfigurationSettings.java
+++ b/src/main/java/net/fabricmc/loom/api/RemapConfigurationSettings.java
@@ -60,6 +60,7 @@ public abstract class RemapConfigurationSettings implements Named {
 		getOnCompileClasspath().finalizeValueOnRead();
 		getOnRuntimeClasspath().finalizeValueOnRead();
 		getPublishingMode().convention(PublishingMode.NONE).finalizeValueOnRead();
+		getApplyDependencyTransforms().convention(defaultDependencyTransforms()).finalizeValueOnRead();
 	}
 
 	@Override
@@ -99,6 +100,11 @@ public abstract class RemapConfigurationSettings implements Named {
 	 * @return the {@link PublishingMode} for the configuration.
 	 */
 	public abstract Property<PublishingMode> getPublishingMode();
+
+	/**
+	 * @return true when dependencies should be evaluated for minecraft jar transforms such as transitive Access Wideners or Injected interfaces.
+	 */
+	public abstract Property<Boolean> getApplyDependencyTransforms();
 
 	public enum PublishingMode {
 		NONE,
@@ -173,5 +179,9 @@ public abstract class RemapConfigurationSettings implements Named {
 	@Internal
 	private NamedDomainObjectProvider<Configuration> getConfigurationByName(String name) {
 		return getProject().getConfigurations().named(name);
+	}
+
+	private Provider<Boolean> defaultDependencyTransforms() {
+		return getSourceSet().map(sourceSet -> sourceSet.getName().equals(SourceSet.MAIN_SOURCE_SET_NAME) || sourceSet.getName().equals("client"));
 	}
 }

--- a/src/main/java/net/fabricmc/loom/configuration/processors/SpecContextImpl.java
+++ b/src/main/java/net/fabricmc/loom/configuration/processors/SpecContextImpl.java
@@ -113,10 +113,12 @@ public record SpecContextImpl(List<FabricModJson> modDependencies, List<FabricMo
 						.map(File::toPath);
 
 		final List<Path> runtimeEntries = extension.getRuntimeRemapConfigurations().stream()
+				.filter(settings -> settings.getApplyDependencyTransforms().get())
 				.flatMap(resolve)
 				.toList();
 
 		return extension.getCompileRemapConfigurations().stream()
+				.filter(settings -> settings.getApplyDependencyTransforms().get())
 				.flatMap(resolve)
 				.filter(runtimeEntries::contains) // Use the intersection of the two configurations.
 				.map(FabricModJsonFactory::createFromZipOptional)

